### ISSUE TITLE
Add support for transfer throttling to rclone

### ISSUE THREAD

**Issue by varunsharma0286** [rclone/rclone#7977]
#### What is your current rclone version (output from `rclone version`)?
rclone v1.68.0-beta.8125.f26d2c6ba

#### What problem are you are trying to solve?
On a machine/server which is actively being used for other tasks, a long running sync/copy operation can make the system unresponsive/slow. 

#### How do you think rclone should be changed to solve that?
I am proposing to add below 3 configs to support this change:
* **--throttle** : Flag to enable the throttling behavior 
* **--load-threshold**: Average load threshold to be used for triggering throttling
* **--max-load-threshold**: Maximum avg load threshold limit. Post this, we will completely throttle the rclone transfers

Throttling will only impact new transfers. Running/In progress transfers will not be impacted.  

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by kapitainsky** [comment]
You can always indirectly control rclone load by:

1. checkers and transfers:
https://rclone.org/docs/#checkers-n
https://rclone.org/docs/#transfers-n

2. Bandwidth limits: 
https://rclone.org/docs/#bwlimit-bandwidth-spec


Or directly by usingyour OS features, e.g.: on Linux `nice` command or utilise [cgroups](https://en.wikipedia.org/wiki/Cgroups)

**Comment by varunsharma0286** [comment]
@kapitainsky Thanks for the response. 

Using checkers and transfers number to control the load might not hold for the long running copy operations. 
For instance, in our use case, the sync/copy operation can take several hours (due to the amount of data). 
We will have to keep the initial values conservatively if we want to use checkers and transfers config.

Also, i am not sure if using `cgroups`(or equivalent) approach would be useful. 
For instance, rclone can be used on a retail machine (e.g. Windows/Macbook). 
For instance, in our use case, the rclone is run on demand on the remote machines as a one time operation. 
These are instances, where we observed the system becoming slow on load. 
I am not sure if its correct to expect the users to configure cgroups.

But i agree that using the load average approach for throttling is platform specific.
One split approach i can think of is:
* Use the CPU+memory as a trigger for throttling in windows. rclone can take the limit as config.
* Use load average for the non-windows systems.

Or alternative would be to use the CPU+memory as a param across all platforms.

Let me know your thoughts on this.

**Comment by MuhammadAbrar1984** [comment]
This makes sense to me. When handling a significant amount of data, I still want my system to remain usable while allowing rclone to utilize maximum resources. Although average load is platform-specific, it is consistent across all Unix-based OSes. On Windows, we can use CPU utilization or even detect user activity. This way, when the user is actively using the system, we can reduce rclone's load, and when the system is idle, rclone can run at maximum capacity. I believe the default copy operation in Windows implements a similar mechanism

**Comment by ncw** [comment]
On linux using `ionice` is quite a good solution for making rclone use less resources.

I do this in the scripts I use rclone in

```
# lower CPU and disk priority
renice 19 $$
ionice -c 3 -p $$
```

Which sets rclone to use only idle CPU and idle IO bandwidth.

Arguably rclone should do this itself, or have an option to do it.

Going back to the original proposal

> am proposing to add below 3 configs to support this change:
> 
> * **--throttle** : Flag to enable the throttling behavior
> * **--load-threshold**: Average load threshold to be used for triggering throttling
> * **--max-load-threshold**: Maximum avg load threshold limit. Post this, we will completely throttle the rclone transfers
> 
> Throttling will only impact new transfers. Running/In progress transfers will not be impacted.

Load average as a concept is Unix cross platform, though it annoyingly inconsistent to read.

| GOOS	| Source |
|---|---|
| darwin | getloadavg() |
| dragonfly | getloadavg() |
| freebsd | /proc/loadavg |
| linux | /proc/loadavg |
| netbsd | getloadavg() |
| openbsd | getloadavg() |

I don't think load average is a particularly good measure of machine load though. It measures CPU well but a machine can be brought to its knees by using too much IO more easily probably.

If we could figure out how to do the `ionice` and `nice` commands above on a cross platform basis then I think that would be a better solution. `nice` will work for all unix OSes. Windows can do IO prioritization I think too.

How about that for a solution? So a flag - say `--low-priority` which uses OS best practices to make rclone run with low IO and CPU priority.

**Comment by MuhammadAbrar1984** [comment]
Yes, determining the average load (avgload) can be complex, but ionice is not expected to function optimally in all configurations. The main challenge lies with the disk scheduler, as ionice primarily works with the CFQ scheduler, if I recall correctly. On the other hand, avgload is a relatively complex formula used to determine system load. Using ionice requires a deeper understanding of what other processes are running and their respective I/O priorities. In contrast, avgload is a straightforward parameter that doesn't require as much detail to understand. It can be used to set a maximum avgload limit, beyond which the system's performance may be compromised.

**Comment by rcfa** [comment]
It would be nice if this weren't restricted to sync, but would also work for mount mode.

Here's the scenario: I have a remote (in my case internxt via its local WebDAV proxy) mounted with rclone as system backup location for a vhost.

The vhost has 100GB storage, of which currently around 40GB are full.

If I do a multi-volume tar backup to the rclone mounted internxt storage, then tar writes files much faster than they can be written to the server, resulting in the vfs cache blowing up in size.

As currently there's more free storage space than there's data to be backed up, this is not pretty, but still works. But imagine the 75GB of the 100GB being used in the future, and the same behavior would result in my vhost running out of disk space.

So being able to limit how fast the vfs mount of a remote can be written to, prevents this problem. One can use a limit near (best: slightly below) the average write speed to the remote) and vfs can always keep up with writing files back to the remote store, without provoking the local cache blowing up in size due to the write backlog.

Currently, I try to force something similar by limiting the cache size, but that's a bit like pushing rope.

**Comment by jrudolph** [comment]
ionice did not have a big enough impact in our case. My guess is that the high amount of concurrency can still starve out other tasks' IO.

cgroups works well however, here's what I did: https://gist.github.com/jrudolph/4ca575f5b9b6505a82545ca71a3badcc

**Comment by scottnzuk** [comment]
guys i found out why it will never work.. internxt don't really have webdav support.
https://github.com/internxt/cli/issues/248

they don't support two most critial features need. Copy function being one of them almost laughable. im not sure if the wizards at RCLONE can make a profile that is not use copy or proppatch.


COPY        | ❌ | Enables in-place duplication, versioning, efficient syncing
PROPPATCH | ❌ | Enables timestamp and metadata preservation

**Comment by scottnzuk** [comment]
guys i found out why it will never work.. internxt don't really have webdav support.
https://github.com/internxt/cli/issues/248

they don't support two most critial features need. Copy function being one of them almost laughable. im not sure if the wizards at RCLONE can make a profile that is not use copy or proppatch.

COPY | ❌ | Enables in-place duplication, versioning, efficient syncing
PROPPATCH | ❌ | Enables timestamp and metadata preservation
